File: rgctl-tests/ecommerce-cpp/src/coolstore/services/shopping_cart_service.cpp
#include "ecommerce/coolstore/services/shopping_cart_service.hpp"

namespace ecommerce::coolstore {

ShoppingCartService::ShoppingCartService(CoolstoreProductService& productService,
                                         PromoService& promoService,
                                         ShippingService& shippingService,
                                         CoolstoreOrderService& orderService)
    : productService_(productService),
      promoService_(promoService),
      shippingService_(shippingService),
      orderService_(orderService) {}

ShoppingCart& ShoppingCartService::getShoppingCart(const std::string& cartId) {
    auto it = carts_.find(cartId);
    if (it == carts_.end()) {
        it = carts_.emplace(cartId, ShoppingCart(cartId)).first;
    }
    return it->second;
}

CatalogProduct* ShoppingCartService::getProduct(const std::string& itemId) {
    return productService_.getProductByItemId(itemId);
}

ShoppingCart& ShoppingCartService::checkOutShoppingCart(const std::string& cartId) {
    ShoppingCart& cart = getShoppingCart(cartId);
    priceShoppingCart(&cart);
    orderService_.process(cart);
    cart.resetShoppingCartItemList();
    priceShoppingCart(&cart);
    return cart;
}

void ShoppingCartService::initShoppingCartForPricing(ShoppingCart& sc) {
    sc.cartItemTotal = 0;
    sc.cartItemPromoSavings = 0;
    sc.shippingTotal = 0;
    sc.shippingPromoSavings = 0;
    sc.cartTotal = 0;

    for (auto& sci : sc.shoppingCartItemList) {
        if (sci.hasProduct) {
            CatalogProduct* p = getProduct(sci.product.itemId);
            if (p) {
                sci.product = *p;
                sci.price = p->price;
            }
        }
        sci.promoSavings = 0;
    }
}

/** Mutates ShoppingCart totals — primary CPG field-write site. */
void ShoppingCartService::priceShoppingCart(ShoppingCart* sc) {
    if (!sc) return;
    initShoppingCartForPricing(*sc);

    if (!sc->shoppingCartItemList.empty()) {
        promoService_.applyCartItemPromotions(*sc);

        for (auto& sci : sc->shoppingCartItemList) {
            sc->cartItemPromoSavings =
                sc->cartItemPromoSavings + sci.promoSavings * sci.quantity;
            sc->cartItemTotal = sc->cartItemTotal + sci.price * sci.quantity;
        }

        sc->shippingTotal = shippingService_.calculateShipping(*sc);
        if (sc->cartItemTotal >= 25) {
            sc->shippingTotal =
                sc->shippingTotal + shippingService_.calculateShippingInsurance(*sc);
        }
    }

    promoService_.applyShippingPromotions(*sc);
    sc->cartTotal = sc->cartItemTotal + sc->shippingTotal;
}

std::vector<ShoppingCartItem> ShoppingCartService::dedupeCartItems(
    const std::vector<ShoppingCartItem>& cartItems) {
    std::unordered_map<std::string, int> quantityMap;
    for (const auto& sci : cartItems) {
        if (!sci.hasProduct) continue;
        quantityMap[sci.product.itemId] += sci.quantity;
    }
    std::vector<ShoppingCartItem> result;
    for (const auto& [itemId, quantity] : quantityMap) {
        CatalogProduct* p = getProduct(itemId);
        if (!p) continue;
        ShoppingCartItem newItem;
        newItem.quantity = quantity;
        newItem.price = p->price;
        newItem.product = *p;
        newItem.hasProduct = true;
        result.push_back(newItem);
    }
    return result;
}

}  // namespace ecommerce::coolstore

File: rgctl-tests/ecommerce-csharp/src/Ecommerce/Coolstore/Services/ShoppingCartService.cs
using System.Collections.Concurrent;
using Ecommerce.Coolstore.Models;

namespace Ecommerce.Coolstore.Services;

public class ShoppingCartService
{
    private readonly CoolstoreProductService _productService;
    private readonly PromoService _promoService;
    private readonly ShippingService _shippingService;
    private readonly CoolstoreOrderService _orderService;
    private readonly ConcurrentDictionary<string, ShoppingCart> _carts = new();

    public ShoppingCartService(
        CoolstoreProductService productService,
        PromoService promoService,
        ShippingService shippingService,
        CoolstoreOrderService orderService)
    {
        _productService = productService;
        _promoService = promoService;
        _shippingService = shippingService;
        _orderService = orderService;
    }

    public ShoppingCart GetShoppingCart(string cartId) =>
        _carts.GetOrAdd(cartId, id => new ShoppingCart(id));

    public CatalogProduct? GetProduct(string itemId) =>
        _productService.GetProductByItemId(itemId);

    public ShoppingCart CheckOutShoppingCart(string cartId)
    {
        var cart = GetShoppingCart(cartId);
        PriceShoppingCart(cart);
        _orderService.Process(cart);
        cart.ResetShoppingCartItemList();
        PriceShoppingCart(cart);
        return cart;
    }

    /** Mutates ShoppingCart totals — primary CPG field-write site. */
    public void PriceShoppingCart(ShoppingCart? sc)
    {
        if (sc == null)
        {
            return;
        }

        InitShoppingCartForPricing(sc);

        if (sc.ShoppingCartItemList.Count > 0)
        {
            _promoService.ApplyCartItemPromotions(sc);

            foreach (var sci in sc.ShoppingCartItemList)
            {
                sc.CartItemPromoSavings += sci.PromoSavings * sci.Quantity;
                sc.CartItemTotal += sci.Price * sci.Quantity;
            }

            sc.ShippingTotal = _shippingService.CalculateShipping(sc);
            if (sc.CartItemTotal >= 25)
            {
                sc.ShippingTotal += _shippingService.CalculateShippingInsurance(sc);
            }
        }

        _promoService.ApplyShippingPromotions(sc);
        sc.CartTotal = sc.CartItemTotal + sc.ShippingTotal;
    }

    private void InitShoppingCartForPricing(ShoppingCart sc)
    {
        sc.CartItemTotal = 0;
        sc.CartItemPromoSavings = 0;
        sc.ShippingTotal = 0;
        sc.ShippingPromoSavings = 0;
        sc.CartTotal = 0;

        foreach (var sci in sc.ShoppingCartItemList)
        {
            if (sci.Product != null)
            {
                var p = GetProduct(sci.Product.ItemId);
                if (p != null)
                {
                    sci.Product = p;
                    sci.Price = p.Price;
                }
            }

            sci.PromoSavings = 0;
        }
    }

    public List<ShoppingCartItem> DedupeCartItems(List<ShoppingCartItem> cartItems)
    {
        var quantityMap = new Dictionary<string, int>();
        foreach (var sci in cartItems)
        {
            if (sci.Product == null)
            {
                continue;
            }

            var itemId = sci.Product.ItemId;
            quantityMap[itemId] = quantityMap.GetValueOrDefault(itemId) + sci.Quantity;
        }

        var result = new List<ShoppingCartItem>();
        foreach (var (itemId, quantity) in quantityMap)
        {
            var p = GetProduct(itemId);
            if (p == null)
            {
                continue;
            }

            result.Add(new ShoppingCartItem
            {
                Quantity = quantity,
                Price = p.Price,
                Product = p
            });
        }

        return result;
    }
}
